src/adapter/src/optimize/dataflows.rs

```diff
@@ -305,9 +304,13 @@ impl<'a> DataflowBuilder<'a> {
 
     /// Determine the given view's monotonicity.
     ///
-    /// This recursively traverses the expressions of all (materialized) views involved in the
+    /// This recursively traverses the expressions of all (non-materialized) views involved in the
     /// given view's query expression. If this becomes a performance problem, we could add the
     /// monotonicity information of views into the catalog instead.
+    ///
+    /// Note that materialized views are never monotonic, no matter their definition, because the
+    /// self-correcting persist_sink may insert retractions to correct the contents of its output
+    /// collection.
     fn monotonic_view(&self, id: GlobalId) -> bool {
```

Comment:
I'm wondering if `import_into_dataflow`'s `MaterializedView` case should also be modified. It's not strictly necessary for correctness, but might make the code more clear.

Reply:
> I did modify it above (setting it to false instead of calling monotonic_view(). Did you have something else in mind?

Ah, sorry, I got confused somehow. (I think I was looking at that function in my IDE, but actually I had something else checked out locally, not this PR.)

>> Or, how about refactoring `import_into_dataflow` to do `let monotonic = self.monotonic_view(*id);` just once at the beginning of the method, and then use that monotonic in every branch?

> We can do that but it does feel a bit strange to call a method named `monotonic_view` on things like sources, tables and MVs. Either the method is intended to return the correct result for any object type, in which case we should rename it to something like `is_monotonic`, or it is intended to only return the correct result for views, in which case we should only use it for views too.

Then maybe we could rename it. It seems that it already returns the correct result for any object type, right?

Reply:
> It seems that it already returns the correct result for any object type, right?

Well, it does intentionally skip indexes, which seems like a pitfall if we don't call it `is_monotonic_and_not_an_index` 😬 Is there a reason why we can't just make it support indexes too?

Reply:
Oops 😬 Well, `import_into_dataflow` wouldn't call it on an index, because it would call it just on its `id` argument, which is not an index. So, `is_monotonic_and_not_an_index` would be kind of ok (with maybe a soft_assert that it's really not an index). Or if we deem that too weird, then we could make it handle the index case also, by looking up the index in the catalog and recursing to the object that the index is on. (The `DataflowBuilder` has a `CatalogState` ref.)

Reply:
Yeah, I think `is_monotonic_and_not_an_index` would be pretty weird :D I've added the second option as a second commit now, lmk what you think.